core/controllers/library_test.py

```diff
@@ -194,14 +197,148 @@ def test_library_handler_for_created_explorations(self):
+        self.signup(self.VIEWER_EMAIL, self.VIEWER_USERNAME)
+        self.viewer_id = self.get_user_id_from_email(self.VIEWER_EMAIL)
+
+    def test_library_index_handler_for_user_preferred_language(self):
+        """Test whether the handler returns the correct language preference."""
+        # Since the default language is 'en', the language preference for the
+        # viewer is changed to 'de' to test if the preference returned is user's
+        # preference.
+        self.login(self.VIEWER_EMAIL)
+        response = self.get_html_response('/preferences')
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        # Change the user's preferred language to de.
+        self.put_json(
+            '/preferenceshandler/data',
+            {'update_type': 'preferred_language_codes', 'data': ['de']},
+            csrf_token=csrf_token)
+        response_dict = self.get_json(feconf.LIBRARY_INDEX_DATA_URL)
+        self.assertDictContainsSubset({
+            'activity_summary_dicts_by_category': [],
+            'preferred_language_codes': ['de'],
+        }, response_dict)
+
+    def test_library_index_handler_for_top_rated(self):
+        """Test the handler for top rated explorations."""
+        response_dict = self.get_json(feconf.LIBRARY_INDEX_DATA_URL)
+        self.assertDictContainsSubset({
+            'activity_summary_dicts_by_category': [],
+            'preferred_language_codes': ['en'],
+        }, response_dict)
+
+        # Load a demo.
+        exp_services.load_demo('0')
+        rating_services.assign_rating_to_exploration('user', '0', 2)
+        response_dict = self.get_json(feconf.LIBRARY_INDEX_DATA_URL)
+        self.assertEqual(
+            len(response_dict['activity_summary_dicts_by_category']), 1)
+        self.assertDictContainsSubset({
+            'preferred_language_codes': ['en'],
+        }, response_dict)
+        activity_summary_dicts_by_category = (
+            response_dict['activity_summary_dicts_by_category'][0])
+        self.assertDictContainsSubset({
+            'categories': [],
+            'header_i18n_id': (
+                feconf.LIBRARY_CATEGORY_TOP_RATED_EXPLORATIONS),
+            'has_full_results_page': True,
+            'full_results_url': feconf.LIBRARY_TOP_RATED_URL,
+            'protractor_id': 'top-rated',
+        }, activity_summary_dicts_by_category)
+        activity_summary_dicts = (
+            activity_summary_dicts_by_category['activity_summary_dicts'])
+        self.assertEqual(
+            len(activity_summary_dicts), 1)
+        self.assertDictContainsSubset({
+            'id': '0',
+            'category': 'Welcome',
+            'title': 'Welcome to Oppia!',
+            'language_code': 'en',
+            'objective': 'become familiar with Oppia\'s capabilities',
+            'status': rights_manager.ACTIVITY_STATUS_PUBLIC,
+        }, activity_summary_dicts[0])
+
+    def test_library_index_handler_for_featured(self):
```

Comment:
Maybe name should be something like `test_library_index_handler_updates_featured_activity_summary_dicts`?

Reply:
renamed